File: Example/MWTabViewExampleObjC/ViewController.m
//
//  ViewController.m
//  MWTabViewExampleObjC
//
//  Created by Clément Padovani on 6/4/16.
//  Copyright © 2016 Mathieu White. All rights reserved.
//

#import "ViewController.h"

@import MWTabView;

@interface ViewController () <MWTabViewDelegate, MWTabViewDataSource>

@property (nonatomic, copy) NSArray <UIView *> *tabViewViews;

@property (nonatomic, weak) MWTabView *tabView;

@end

@implementation ViewController

- (void)viewDidLoad
{
    [super viewDidLoad];

    NSMutableArray <UIView *> *tabViewViews = [NSMutableArray arrayWithCapacity: 3];

    NSArray <UIColor *> *colors = @[[UIColor redColor],
                                    [UIColor greenColor],
                                    [UIColor blueColor]];

    for (UIColor *aColor in colors)
    {
        UIView *aView = [[UIView alloc] init];

        [aView setBackgroundColor: aColor];

        [tabViewViews addObject: aView];
    }

    [self setTabViewViews: tabViewViews];

    MWTabView *tabView = [[MWTabView alloc] init];

    [tabView setBackgroundColor: [UIColor lightGrayColor]];

    NSArray <UIColor *> *gradientColors = @[[UIColor whiteColor],
                                            [UIColor colorWithWhite: .9f alpha: 1]];

    [tabView setGradientForTabViewHeaderWithColors: gradientColors];

    [tabView setDelegate: self];

    [tabView setDataSource: self];

    [tabView setTranslatesAutoresizingMaskIntoConstraints: NO];

    [[self view] addSubview: tabView];

    [self setTabView: tabView];

    [self setupConstraints];
}

- (void) setupConstraints
{
    NSDictionary *viewsDictionary = NSDictionaryOfVariableBindings(_tabView);

    NSDictionary *metricsDictionary = nil;

    [[self view] addConstraints: [NSLayoutConstraint constraintsWithVisualFormat: @"H:|[_tabView]|"
                                                                         options: 0
                                                                         metrics: metricsDictionary
                                                                           views: viewsDictionary]];

    [[self view] addConstraints: [NSLayoutConstraint constraintsWithVisualFormat: @"V:|[_tabView]|"
                                                                         options: 0
                                                                         metrics: metricsDictionary
                                                                           views: viewsDictionary]];
}

- (NSUInteger) numberOfTabsInTabView: (MWTabView *) tabView
{
    return 3;
}

- (UIView *) tabView: (MWTabView *) tabView viewForTabAtIndex: (NSUInteger) index
{
    return [self tabViewViews][index];
}

- (NSString *) tabView: (MWTabView *) tabView titleForTabAtIndex: (NSUInteger) index
{
    return [NSString stringWithFormat: @"My Tab %lu", index + 1];
}

// MARK: - MWTabViewDelegate Methods

- (CGFloat) heightForHeaderInTabView: (MWTabView *) tabView
{
    return 64;
}

- (CGFloat) widthForTitlesInTabView: (MWTabView *) tabView
{
    return ceilf(CGRectGetWidth([[self view] bounds]) / 3.f);
}

- (BOOL) tabViewShouldRecycleViews: (MWTabView *) tabView
{
    return YES;
}

@end

File: Example/MWTabViewExample/ViewController.swift
//
//  ViewController.swift
//  MWTabViewExample
//
//  Created by Mathieu White on 2016-06-04.
//  Copyright © 2016 Mathieu White. All rights reserved.
//

import UIKit

import MWTabView

class ViewController: UIViewController, MWTabViewDataSource, MWTabViewDelegate {
    
    
    // MARK: - Variables
    
    /// The tab view for the example.
    weak var tabView: MWTabView?
    
    
    // MARK: - Initialization
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        // Initialize the tab view
        let tabView = MWTabView()
        tabView.backgroundColor = UIColor.lightGrayColor()
        tabView.setGradientForTabViewHeader(colors: [UIColor.whiteColor(), UIColor(white: 0.9, alpha: 1.0)])
        tabView.translatesAutoresizingMaskIntoConstraints = false
        tabView.dataSource = self
        tabView.delegate = self
        
        // Add the tab view to the main view
        self.view.addSubview(tabView)
        
        // Set the tab view to its variable
        self.tabView = tabView
        
        // Auto Layout
        self.setupConstraints()
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    
    // MARK: - Auto Layout

    private func setupConstraints() {
        // Views Dictionary
        var viewsDict: [String : AnyObject] = [String : AnyObject]()
        viewsDict["_tabView"] = self.tabView
        
        // Tab View Vertical Constraints
        self.view.addConstraints(NSLayoutConstraint.constraintsWithVisualFormat("V:|[_tabView]|",
            options: NSLayoutFormatOptions(rawValue: 0),
            metrics: nil,
            views: viewsDict))
        
        // Tab View Horizontal Constraints
        self.view.addConstraints(NSLayoutConstraint.constraintsWithVisualFormat("H:|[_tabView]|",
            options: NSLayoutFormatOptions(rawValue: 0),
            metrics: nil,
            views: viewsDict))
    }
    
    
    // MARK: - MWTabViewDataSource Methods
    
    func numberOfTabsInTabView(tabView: MWTabView) -> UInt {
        return 3
    }
    
    func tabView(tabView: MWTabView, viewForTabAtIndex index: UInt) -> UIView {
        let view = UIView()
        
        switch (index) {
        case 0: view.backgroundColor = UIColor.redColor()
        case 1: view.backgroundColor = UIColor.greenColor()
        case 2: view.backgroundColor = UIColor.blueColor()
        default: view.backgroundColor = UIColor.whiteColor()
        }
        
        return view
    }

    func tabView(tabView: MWTabView, titleForTabAtIndex index: UInt) -> String {
        return String(format: "My Tab %lu", index + 1)
    }
    
    
    // MARK: - MWTabViewDelegate Methods
    
    func heightForHeaderInTabView(tabView: MWTabView) -> CGFloat {
        return 64.0
    }
    
    func widthForTitlesInTabView(tabView: MWTabView) -> CGFloat {
        return ceil(CGRectGetWidth(self.view.bounds) / 3.0)
    }
    
    func tabViewShouldRecycleViews(tabView: MWTabView) -> Bool {
        return true
    }

}

